src/components/Collapsible/CollapsibleHooks.ts

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react'
+
+interface UseCollapsible {
+	ref: React.RefObject<HTMLDivElement>
+}
+
+interface UseCollapsibleProps {
+	collapsed: boolean
+}
+
+export const useCollapsible = ({
+	collapsed
+}: UseCollapsibleProps): UseCollapsible => {
+	const ref = React.useRef<HTMLDivElement>(null)
+	const hasMounted = React.useRef(false)
+	const onTransitionEnd = React.useRef<() => void>()
+
+	React.useLayoutEffect(() => {
+		const onCollapsed = (): void => {
+			if (ref.current == null) return
+
+			if (!hasMounted.current) {
+				ref.current.style.height = '0px'
+				ref.current.style.visibility = 'hidden'
+				hasMounted.current = true
+				return
+			}
+
+			const height = ref.current.scrollHeight
+
+			onTransitionEnd.current = (): void => {
+				if (ref.current == null) return
+
+				if (onTransitionEnd.current != null) {
+					ref.current.removeEventListener(
+						'transitionend',
+						onTransitionEnd.current
+					)
+				}
+
+				ref.current.style.visibility = 'hidden'
+			}
+
+			if (onTransitionEnd.current != null) {
+				ref.current.addEventListener('transitionend', onTransitionEnd.current)
+			}
+
+			requestAnimationFrame(() => {
+				if (ref.current == null) return
+
+				ref.current.style.height = `${height}px`
+
+				requestAnimationFrame(() => {
+					if (ref.current == null) return
+
+					ref.current.style.height = '0px'
+				})
+			})
+		}
+
+		const onExpanded = (): void => {
+			if (ref.current == null) return
+
+			ref.current.style.visibility = 'visible'
+
+			if (!hasMounted.current) {
+				ref.current.style.height = ''
+				hasMounted.current = true
+				return
+			}
+
+			onTransitionEnd.current = (): void => {
+				if (ref.current == null) return
+
+				if (onTransitionEnd.current != null) {
+					ref.current.removeEventListener(
```

Comment:
Wow, man, I don't know. It seems to me like a lot of code :-) I think that all the parts including visibility are completely unnecessary (lines 24, 31-42 and 64). Overflow hidden and setting the height must be enough. Also on the line 83, setting overflow to empty is not necessary because you don't set it anywhere. And I think that this removeEventListener is not working (line 76). Try to put console.log here and you'll see that for the first time, it's called correctly on expansion but after the first go, it's called every time, also on collapse, which is incorrect, I believe.

I mean, I know that it works :-D but let's try to optimise it a bit. Or I don't, know maybe it's necessary there and I'm missing something. Let me know, what you think.

Reply:
I will have a look at simplifying this

Reply:
If we remove it from DOM, we won't need that visibility for tests, right? I have to say it's a bit too complex for me to say just from my head what the best solution will be. But I have a feeling that maybe completely removing it by setting props and re-rendering could result in simpler and more readable code. Hard to say :-/ I'd give it a go though.